Feat/hand limit rework (#1359)
cuttle-cards/cuttle: docs/game-state-api.md

Author: I meant to add a skill instructing the agent to read this doc when trying to understand the backend architecture and request lifecycle

# Move History and Gamestate API

> This doc outlines the database, backend processing, and client socket payload architecture of processing individual moves and their resulting game states in cuttle.cards. This powers the core game loop and facilitates a variety of features that leverage granular data collection and decouple the representation of where cards are located from the representation of the users so a given user can be in more than one game at once. In particular makes possible:

Author: ```suggestion
This doc outlines the database, backend processing, and client socket payload architecture of processing individual moves and their resulting game states in cuttle.cards. This powers the core game loop and facilitates a variety of features that leverage granular data collection. In particular makes possible:
```

* Building the vs AI experience into the main application   

> * Watchable game replayed of completed games   

Author: ```suggestion
* Watchable game replays of completed games   
```

* Async play
* Strategic analysis of gameplay

# Enums

## GamePhase

Enum describing the phase of a turn the game is currently in. Used to validate which next-moves are legal.

* main \- 'main' player phase. Can make 'regular' moves:  
   * draw  
   * points  
   * faceCard  
   * jack  
   * scuttle  
   * untargetedOneOff  

>    * targetedOneOff  

Author: Add requestStalemate

Author: ```suggestion
   * targetedOneOff  
   * requestStalemate
   
```

* countering \- phase where players play counters. Only legal moves are   
   * counter  
   * resolve  
* resolvingThree \- Picking card from scrap. Only legal move is resolveThree  

> * resolvingFour \- Choosing cards to discard. Only legal move is resolveFour  

Author: ```suggestion
* resolvingFour \- Choosing cards to discard due to opponent four. Only legal move is resolveFour  
```

* resolvingFive \- Choosing card to discard before drawing. Only legal move is resolveFive  
* resolvingSeven \- Picking one of the top cards from the deck. Must play a seven move next. Legal next moves:  
   * sevenPoints  
   * sevenFaceCard  
   * sevenScuttle  
   * sevenUntargetedOneOff  
   * sevenTargetedOneOff  
   * sevenJack  
* discardingToHandLimit \- discarding due to hand limit at end of turn. Only legal move is discardToHandLimit  
* consideringStalemate \- deciding whether to accept opponent's stalemate request. Only legal moves are stalemateAccept and stalemateReject

## MoveType

Enum designating which kind of move was made.

* initialize  
* draw  
* points  
* scuttle  
* faceCard (king, queen, or glasses eight)  
* jack  
* untargetedOneOff  
* targetedOneOff  
* counter  
* resolve  
* resolveThree (picking a card from the scrap)  
* resolveFour (discarding from hand)  
* sevenPoints  
* sevenScuttle  
* sevenFaceCard  
* sevenJack  
* sevenUntargetedOneOff  
* sevenTargetedOneOff  
* pass  
* discardToHandLimit  
* stalemateRequest  
* stalemateReject  
* stalemateAccept

# Database Layer

## Game

> The `Game` object/table is reduced to the metadata about the game that doesn’t generally change while a game is played. It has the game’s id, name, the ids of which users are p0 and p1.

Author: ```suggestion
The `Game` object/table contains the metadata about the game that doesn’t generally change while a game is played. It has the game’s id, name, the ids of which users are p0 and p1.
```

* id: primary key  
* playedBy: `int` 0 | 1: Which player made the move (0 if p0, 1 if p1)  
* moveType: `MoveType` \- designates which kind of move was made  
* turn: `int` \- which turn number the move was made on  
* phase: `GamePhase` \- What phase of a turn the game is currently in. Used to validate which next-moves are legal  
* playedCard: `String | null` the card that was played  
* targetCard: `String | null` the card that was targeted  
* discardedCards: `Array<String>`  
* p0Hand: `Array<String>`  
   * Specially formatted string representing the list of cards in p0’s hand. See [‘Game State Array\<String\> Lists](#game-state-array\<string\>-lists) below for full explanation  
   * ex) `[‘AC’, ‘5H’, ‘8D(JS-p0,JD-p1,JH-p0)’]`  
      * P0 hand has the Ace of Clubs, 5 of Hearts, 8 of Diamonds  
         * 8 of diamonds has the jack of spades, jack of diamonds, and jack of hearts on it  
* p1Hand: `Array<String>`: Cards in p1 hand  
* p0Points: `Array<String>`: Cards in p0 points  
* p1Points: `Array<String>`  
* p0FaceCards: `Array<String>`  
* p1FaceCards: `Array<String>`  

> * deck: `Array<String>`: Cards in the deck, in order (this removes the need for topCard and secondCard)  

Author: ```suggestion
* deck: `Array<String>`: Cards in the deck, in order
```

> * scrap: `Array<String>`  

Author: ```suggestion
* scrap: `Array<String>`: list of cards currently in the scrap
```

> * oneOff: `String | null`  

Author: ```suggestion
* oneOff: `String | null`: the one-off currently in play, before resolution. Unset once resolved
```

> * oneOffTarget: `String | null`  

Author: ```suggestion
* oneOffTarget: `String | null`: target of the currently resolving one-off. Unset once one-off resolves 
```

> * twos: `Array<String>`  

Author: ```suggestion
* twos: `Array<String>`: array of twos played as counters, currently on the stack, awaiting resolution
```

> * resolving: `String`  

Author: ```suggestion
* resolving: `String`: one-off being resolved in a secondary step e.g. a three while awaiting player selection from the scrap
```

   /\*\*  
    \* @param {Object} requestedMove \- Object describing the request to draw  
    \* @param {1 | 0} requestedMove.playedBy \- Which player is drawing  
    \* @param { number } requestedMove.turn \- Turn number (should be 1 greater than currentState.turn)  
    \* @param { MoveType.DRAW } requestedMove.moveType \- Specifies that this a draw move  
    \*/  
   requestedMove: {  
     type: 'ref',  
     description: 'The move being requested. Specifies which player is asking to draw a card',  
   }  
  },  
  sync: true, // synchronous helper  
  fn: ({ currentState, requestedMove, playedBy }, exits) => {
    let result = _.cloneDeep(currentState);

    const player = playedBy ? result.p1 : result.p0;

    player.hand.push(result.deck.shift());
    const playerMustDiscard = player.hand.length > 8;

    result = {
      ...result,
      ...requestedMove,
      phase: playerMustDiscard ? GamePhase.DISCARDING_TO_HAND_LIMIT : GamePhase.MAIN,
      turn: playerMustDiscard ? result.turn : result.turn + 1,
      playedBy,
      playedCard: null,
      targetCard: null,
      discardedCards: [],
      resolved: null,
    };

    return exits.success(result);
  },
};
```

> The `publishGameState()` helper is used to send the latest game state to the client. For the MVP rollout, it will emit an event that matches the current data structure used by the client in production, e.g:

Author: ```suggestion
The `publishGameState()` helper is used to send the latest game state to the client
```

```javascript
     Game.publish(\[fullGame.id\], {  
       change: 'points',  
       game: fullGame,  
       victory,  
     });
```

> The published event should have the following shape:  

Author: ```suggestion
The published event has the following shape:  
```

# Move History and Gamestate API

This doc outlines a proposal for the database and api payload architecture of individual moves and the resulting game states in cuttle.cards. This is intended to facilitate a variety of features that leverage more granular data collection (ie the game state after each move) and decouple the representation of where cards are located from the representation of the users so a given user can be in more than one game at once. In particular this will make possible:

* Building the vs AI experience into the main application   
* Watchable game replayed of completed games ✅  
* Async play  

> * Strategic analysis of gameplay

Author: Rephrase as it's no longer rec

   12. sevenPoints  
   13. sevenScuttle  
   14. sevenFaceCard  
   15. sevenJack  
   16. sevenUntargetedOneOff  
   17. sevenTargetedOneOff  
   18. pass  
4. turn: `int` \- which turn number the move was made on  
5. phase: `enum` \- What phase of a turn the game is currently in. Used to validate which next-moves are legal  
   1. main \- ‘main’ player phase. Can make ‘regular’ moves:  
      1. draw  
      2. points  
      3. faceCard  
      4. jack  
      5. scuttle  
      6. untargetedOneOff  
      7. targetedOneOff  
   2. countering \- phase where players play counters. Only legal moves are   
      1. counter  
      2. resolve  
   3. resolvingThree \- Picking card from scrap. Only legal move is resolveThree  
   4. resolvingFour \- Choosing cards to discard. Only legal move is resolveFour  
   5. resolvingFive \- Choosing card to discard before drawing. Only legal move is resolveFive  
   6. resolvingSeven \- Picking one of the top cards from the deck. Must play a seven move next. Legal next moves:  
      1. sevenPoints  
      2. sevenFaceCard  
      3. sevenScuttle  
      4. sevenUntargetedOneOff  
      5. sevenTargetedOneOff  

>       6. sevenJack  

Author: Add new discarding to hand limit phase

6. playedCard: `String | null` the card that was played  

>    1. Can be null  

Author: Don't need these since it's in the type

`p0Points: [‘3D’, ‘4S(JS-p0)’, ‘TH(JH-p0,JC-p1,JD-p0)’]`  
Indicates that player 0 has 3 point cards:

* Three of Diamonds  
* Four of Spades  
  * Has the Jack of Spades (controlled by p0) attached to it  
* Ten of Hearts  
  * Jack of Hearts (p0’s-bottom jack)  
  * Jack of Clubs (p1’s)  
  * Jack of Diamonds (p0’s-top jack)

Jacks are listed from bottom to top order, meaning that the last jack in the list is always the last one played and the one that determines who ultimately controls the point card.

# Backend Processing Layer

## Card

Represents a single card, in its object format.

1. id: `String` (same format as the gameStateRow cards e.g. ‘AS’ for Ace of Spades)  
2. suit: `0 | 1 | 2 | 3`  
3. rank: `[1- 13]`  
4. isFrozen: `Boolean` \- whether the card is currently frozen due to a 9 and can’t be played this turn

## Player

Represents one player in the game.

1. hand: `Array<Card>`  
2. points: `Array<Card>`  
3. faceCards: `Array<Card>`

## GameState

> An uncompressed, object-oriented representation of a `GameStateRow` using the `saveGameState()` helper. Conversely, a `GameStateRow` can be converted to a `GameState` object using the `unpackGameState()` helper. Generally, game logic is processed using `GameState` objects and then the result is converted back to a `GameStateRow` to be persisted in the database.

Author: Typo; uses unpackGameState()

7. turn: `int` \- which turn number the move was made on  
8. phase: `enum` \- What phase of a turn the game is currently in. Used to validate which next-moves are legal  
   1. main \- ‘main’ player phase. Can make ‘regular’ moves:  
      1. draw  
      2. points  
      3. faceCard  
      4. jack  
      5. scuttle  
      6. untargetedOneOff  
      7. targetedOneOff  
   2. countering \- phase where players play counters. Only legal moves are   
      1. counter  
      2. resolve  
   3. resolvingThree \- Picking card from scrap. Only legal move is resolveThree  
   4. resolvingFour \- Choosing cards to discard. Only legal move is resolveFour  
   5. resolvingFive \- Choosing card to discard before drawing. Only legal move is resolveFive  
   6. resolvingSeven \- Picking one of the top cards from the deck. Must play a seven move next. Legal next moves:  
      1. sevenPoints  
      2. sevenFaceCard  
      3. sevenScuttle  
      4. sevenUntargetedOneOff  
      5. sevenTargetedOneOff  

>       6. sevenJack  

Author: Add new discarding to hand limit phase

14. oneOffTarget: `Card | null` \- when a 2 or 9 one-off is played, this describes which card is targeted by its effect  
15. oneOffTargetType `‘point’ | ‘faceCard’ | ‘jack’`  
    1. When the current oneOff is a 2/9, this describes where the target is located e.g. if the target is the 8 of hearts, is that a point card or face card (glasses)  
16. twos: `Array<Card>` \- array of twos that have been played to counter the current oneOff  
17. resolved: `Card | null` \- for a MoveType.RESOLVE move, this describes which oneOff just resolved or fizzled. This is not considered a “place” the cards can exist as the card is in the scrap at that point. Instead it’s just a description of which oneOff just resolved/fizzled.  
18. gameId: ID \- FK to the games table  
19. createdAt: `timestampz` \- When the move took place

# Api Layer

## Socket Payload

> 1. moveType: enum  

Author: Add phase key/val

   10. oneOffTargetType: enum  
       1. PLAYER  
       2. POINT\_CARD  
       3. FACE\_CARD  
       4. JACK  
   11. attachedToTarget: Card | null  
   12. twos : Array\<Card\>  
   13. turn: int  
   14. log  
       1. array of moves  
   15. spectatingUsers  
       1. array of users  
8. \<Player\>  
   1. Hand Array\<Card\>  
   2. Points Array\<Card\>  
   3. Face Cards Array\<Card\>  
   4. Name String  
   5. ID Int  
9. \<Card\>

   1. id: Int  
   2. suit: int  
   3. rank: int  
   4. Frozen: Bool

# Backend Control Flow

Generally, requests to make moves on the game will lock the requested game to prevent other requests from updating it, retrieve the latest game state, validate the requested move, make the requested changes, persist the changes in the database, and emit a socket event with the resulting state. This is done via a single endpoint in `api/controllers/game/move`. It delegates to move-specific helpers based on the `moveType` specified in the request body. These `execute()` and `validate()` helpers for each move live in a folder named after the move inside `api/helpers/gamestate/moves` e.g. `api/helpers/gamestate/moves/draw/execute.js`

> // api/controller/game/move  

Author: Wrap in code block

>  // Request to make a move  

Author: Whitespace

   const gameState \= unpackGamestate(game.gameStates.at(\-1));  
   if (\!game.gameStates.length || \!gameState) {  
     throw new Error({ message: 'Game has not yet started' });  
   }

   // Verify whether user is in requested game and as which player  
   let playedBy;  
   switch (req.session.usr) {  
     case game.p0.id:  
       playedBy \= 0;  
       break;  
     case game.p1.id:  
       playedBy \= 1;  
       break;  
     default:  
       throw new Error('You are not a player in this game\!');  
   }

   validate(gameState, req.body, playedBy);  
   const updatedState \= execute(gameState, req.body, playedBy);  
   const gameStateRow \= await saveGamestate(updatedState);  
   game.gameStates.push(gameStateRow);  
   await publishGameState(game, updatedState);  
   await sails.helpers.unlockGame(game.lock);

   return res.ok();  
 } catch (err) {  
   //unlock game if failing due to validation  
   if (game?.lock) {  
     try {  
       await sails.helpers.unlockGame(game.lock);  
     } catch (err) {  
       //fall through for generic error handling  
     }  
   }  
   return res.badRequest({ message: err.message });  
 }  
};

> The `validate` helper functions (e.g. `sails.helpers.gamestate.move.draw.validate()`) will be MoveType-specific utilities that throw errors if the move is illegal e.g. because it is not your turn or because another move e.g. countering is ongoing.  

Author: Replace with updated version

     // Only allowed phase  
     case GamePhase.MAIN:  
       break;  
     case GamePhase.COUNTERING:  
       return exits.error({ message: 'Cannot draw while players are countering one-offs' });  
     case GamePhase.RESOLVING\_THREE:  
       return exits.error({ message: 'Cannot draw while player is choosing card from scrap' });  
     case GamePhase.RESOLVE\_FOUR:  
       return exits.error({ message: 'Cannot draw while player chooses what to discard from 4'});  
     case GamePhase.RESOLVE\_FIVE:  
       return exits.error({ message: 'Cannot draw while player chooses what to discard for 5' });  
     case GamePhase.RESOLVING\_SEVEN:  
       return exits.error({ message: 'Cannot draw while someone is playing from the top of the deck' });  
     default:  
       return exits.error({ message: \`Cannot draw because game is corrupted. Unidentified phase: ${currentState.phase}\` });  
   }

   if (requestedMove.playedBy \!== (currentState.turn) % 2) {  
     return exits.error({message: "It's not your turn"});  
   }

   if (currentState.deck.length \< 1) {  
     return exits.error({message: 'No cards in deck'});  
   }

   return exits.success();  
 },  
};

> The `execute()` helpers return a new GameState object that reflects the changes of a given move, including moving cards around, updating the turn, setting the phase, and which player played the move.

Author: Replace with updated version and wrap in code block

 description:  
   'Returns new GameState resulting from requested draw move',

 inputs: {  
   currentState: {  
     type: 'ref',  
     description: 'The latest game state before the requesting player draws a card',  
     required: true,  
   },  
   /\*\*  
    \* @param {Object} requestedMove \- Object describing the request to draw  
    \* @param {1 | 0} requestedMove.playedBy \- Which player is drawing  
    \* @param { number } requestedMove.turn \- Turn number (should be 1 greater than currentState.turn)  
    \* @param { MoveType.DRAW } requestedMove.moveType \- Specifies that this a draw move  
    \*/  
   requestedMove: {  
     type: 'ref',  
     description: 'The move being requested. Specifies which player is asking to draw a card',  
   }  
 },  
 sync: true, // synchronous helper  
 fn: ({ currentState, requestedMove }, exits) \=\> {  
   const result \= \_.cloneDeep(currentState);  
   const player \= requestedMove.playedBy \=== 0 ? result.p0 : result.p1;

   // Move cards  
   const topCard \= result.deck.shift();  
   player.hand.push(topCard);

   result \= {  
     ...result,  
     ...requestedMove,  
   };

   return exits.success(result);  
 },  
};

> The `emitGameState()` helper is used to send the latest game state to the client. For the MVP rollout, it will emit an event that matches the current data structure used by the client in production, e.g:

Author: Rename publishGameState
